Fix WAL file removing on mirrors in archive_mode `always`

InnerLife0 opened the pull request:
Far ago, we've backported (fa759d1bf70c020f884f05cf7ae15c6b9b47b4b1) archive_mode `always` to 6x. Since that, it was tested only manually.
Later, there was a bug [found](https://github.com/greenplum-db/gpdb/pull/14189) around WAL recycling depends on archive_mode. Upstream backported a fix (4a5a95b3a62dadd73af73b76f8fed9deeea842a6), but the condition for `XLogArchivingAlways()` was not backported, because upstream doesn't support it. As a result, WAL files on mirrors may be considered as removable and can be actually removed during restart point. Backporting of 78ea8b5daab9237fd42d7a8a836c1c451765499f solves the problem, but new tap-test was also added to prove it works and to show such cases.
_____
The patch reverts code changes made by a commit backported by upstream. However, upstream commit contains test files which we want to keep.
_____
My idea is to not squash commits during merge keeping revert and backported commits as different entities.

michaelpq committed:
Fix WAL recycling on standbys depending on archive_mode

A restart point or a checkpoint recycling WAL segments treats segments
marked with neither ".done" (archiving is done) or ".ready" (segment is
ready to be archived) in archive_status the same way for archive_mode
being "on" or "always".  While for a primary this is fine, a standby
running a restart point with archive_mode = on would try to mark such a
segment as ready for archiving, which is something that will never
happen except after the standby is promoted.

Note that this problem applies only to WAL segments coming from the
local pg_wal the first time archive recovery is run.  Segments part of a
self-contained base backup are the most common case where this could
happen, however even in this case normally the .done markers would be
most likely part of the backup.  Segments recovered from an archive are
marked as .ready or .done by the startup process, and segments finished
streaming are marked as such by the WAL receiver, so they are handled
already.

Reported-by: Haruka Takatsuka
Author: Michael Paquier
Discussion: https://postgr.es/m/15402-a453c90ed4cf88b2@postgresql.org
Backpatch-through: 9.5, where archive_mode = always has been added.

Cherry-picked from 78ea8b5daab9237fd42d7a8a836c1c451765499f

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65720

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1117961

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1117963

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1117956

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1121122

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1180398

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1180399

Commit:
Revert "Ignore XLOG archive check when creating restart point on mirror segment"

This partially reverts commit 4a5a95b3a62dadd73af73b76f8fed9deeea842a6.
The fix to src/backend/access/transam/xlogarchive.c was incomplete.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67061

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1184710

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1184711

whitehawk commented:
Please fix the description according to checklist from wiki

RekGRpth commented:
> Please fix the description according to checklist from wiki

I think that this is not necessary, because the resulting patch will still consist of three different commits, each with its own description, and will go to the merge.

However, I suggest that in the second commit message you explicitly indicate the hash of which commit it was a backport.

whitehawk commented:
> I think that this is not necessary, because the resulting patch will still consist of three different commits, each with its own description, and will go to the merge.

In this case, if you are going to integrate the commits below 'as is' with their own description:
https://github.com/arenadata/gpdb/commit/51bfdec5e581b2b51fe975bb025bfe2942478058
https://github.com/arenadata/gpdb/commit/5084340492b5a24de888631fefc1e7f204dc6f66
https://github.com/arenadata/gpdb/commit/84897d62654a100abdc22744815f1349afc5e7d7

each description should comply with the checklist. In 84897d62654a100abdc22744815f1349afc5e7d7 the alignment is a bit strange - there are many empty spaces that can be filled.

whitehawk commented on src/test/recovery/t/139_archive_mode_always.pl:
```diff
@@ -0,0 +1,82 @@
+	has_archiving    => 1,
+	allows_streaming => 1);
+my $backup_name = 'my_backup';
+my $master_connstr = $node_master->connstr;
```
I do not see where $master_connstr is used. What is the purpose of it?

InnerLife0 replied:
True. This variable is redundant.

whitehawk commented on src/test/recovery/t/139_archive_mode_always.pl:
```diff
@@ -0,0 +1,82 @@
+# The idea of the test is to check that not archived WAL files are not removed
+# by creation of checkpoint (on master) or restartpoint (on standby).
+
+# Initialize master node, doing archives
```
Mixing continious and non-continious forms of verbs sounds a bit awkward to me. I would suggest to re-phrase, for ex. "Initialize master node, doing archives" -> "Initialize master node and make archives", or smth similar

InnerLife0 replied:
1. Sorry, but I can't find any rule about it. If it's only a personal opinion, then I will ignore it, as none of us is native. If it's a rule, then please give a link.
2. The original comment was left by Alvaro Herrera in his [patch](https://github.com/postgres/postgres/commit/49148645f7f30f461b09618b5342b37425f33b22#diff-b57990e9ef44c06b531f9dd2279d95e1a344348d3298f81820a4eb4690fcb205R9) to PG. He, the author and other reviewers are probably not natives too, but all of them are OK with this.
3. (Deep) IMO, but we shouldn't care SO much about such things here or in any other patch (doesn't matter whether we made this changes or not).

whitehawk commented on src/test/recovery/t/139_archive_mode_always.pl:
```diff
@@ -0,0 +1,82 @@
+# Take backup for standby
+$node_master->backup($backup_name);
+
+# Initialize standby node from backup, fetching WAL from archives
```
The same as above - "Mixing continious and non-continious forms of verbs"

InnerLife0 replied:
The same as above.

whitehawk commented on src/test/recovery/t/139_archive_mode_always.pl:
```diff
@@ -0,0 +1,82 @@
+
+# Switch WAL and do checkpoint. Before version 11, Postgres needs files for
+# two points, so do checkpoint twice (see CreateCheckPoint() for more).
+# Checkpoint requested with CHECKPOINT_WAIT, so it's enough to wait for command
```
"Checkpoint requested" -> "Checkpoint is requested" ? I suppose passive voice was meant

InnerLife0 replied:
Will fix this. Thanks.

InnerLife0 commented:
@whitehawk Can you push all comments all at once?
Pushing them early makes no sense and looks like a spam.

InnerLife0 commented:
> each description should comply with the checklist. In [84897d6](https://github.com/arenadata/gpdb/commit/84897d62654a100abdc22744815f1349afc5e7d7) the alignment is a bit strange - there are many empty spaces that can be filled.

What tool are you using to check commit message?

![image](https://github.com/arenadata/gpdb/assets/42867699/5438d649-6961-4ad3-a10f-b76e47e8e69b)

whitehawk commented:
> What tool are you using to check commit message?

I've checked on web 
![image](https://github.com/arenadata/gpdb/assets/167234/788078b4-10d7-4960-89b1-6178505d186c)

I see, it spoiled the view.

RekGRpth commented:
> > each description should comply with the checklist. In [84897d6](https://github.com/arenadata/gpdb/commit/84897d62654a100abdc22744815f1349afc5e7d7) the alignment is a bit strange - there are many empty spaces that can be filled.
> 
> What tool are you using to check commit message?
> 
> ![image](https://private-user-images.githubusercontent.com/42867699/315911351-5438d649-6961-4ad3-a10f-b76e47e8e69b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwOTM3MDAsIm5iZiI6MTcxMTA5MzQwMCwicGF0aCI6Ii80Mjg2NzY5OS8zMTU5MTEzNTEtNTQzOGQ2NDktNjk2MS00YWQzLWExMGYtYjc2ZTQ3ZThlNjliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDA3NDMyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmYTQyYTkyZDRhZGJlMGFhODJlNzE0NDU0YmZkZTZjOTdiNjI4ZDBiMGRlZDYzMTI3ZThkMzM2MmNmNmVkNGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ObIRaNfHvjoaPFo9fOR6LuuNNKHIvyaTLefDI8Ajp0g)

Reduce commit hashes to the minimum value in the commit message.

InnerLife0 commented:
> Reduce commit hashes to the minimum value in the commit message.

Is it a rule? Some time ago I saw examples of same short hashes for different commits. Github made incorrect link for this.

Commit:
Add tap-test for WAL files in archive_mode `always`

The commit adds tap-test to check that WAL files are not deleted at checkpoint.

Far ago, we've backported (fa759d1bf70c020f884f05cf7ae15c6b9b47b4b1)
archive_mode `always` to 6x. Since that, it was tested only manually.
Later, there was a bug around WAL recycling depends on archive_mode. Upstream
backported a fix (4a5a95b3a62dadd73af73b76f8fed9deeea842a6), but the condition
for `XLogArchivingAlways()` was not backported, because upstream doesn't support
it. As a result, WAL files on mirrors may be considered as removable and can be
actually removed during restart point. Backporting of
78ea8b5daab9237fd42d7a8a836c1c451765499f solves the problem, but new tap-test
was also added to prove it works and to show such cases.

The upstream backport already covers WAL files on master instance, but I decided
to add this check to test file too. Test file is heavily inspired by other test
files from `recovery` dir.

InnerLife0 commented:
The force push was done as commit message of second commit was changed. Now it contains a link to original commit.
I believe we shouldn't divide a head of first commit message, as it's done by revert automatically.
The message of third commit was also changed closer to PR description.
Mistakes in test file were fixed too.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67222

whitehawk commented:
I would suggest rephrasing the commit description to avoid speaking in the first person (e.g. "we've backported", "I decided", etc). 
But as it is not in the checklist , it is up to author for now.

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193313

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193314

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4965

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67427

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204844

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204845
